python/cuml/neighbors/nearest_neighbors.pyx

```diff
@@ -392,6 +433,170 @@ class NearestNeighbors(Base):
         )
 
         self.handle.sync()
+        
+        if not _output_cumlarray:
+            I_ndarr = I_ndarr.to_output(out_type)
+            if return_distance:
+                D_ndarr = D_ndarr.to_output(out_type)
+
+        # drop first column if using training data as X
+        # this will need to be moved to the C++ layer (cuml issue #2562)
+        if use_training_data:
+            if out_type == 'cupy' or out_type == 'numpy' \
+                or out_type == 'numba' or out_type == 'cumlarray':
+                return (D_ndarr[:, 1:], I_ndarr[:, 1:]) \
+                    if return_distance else I_ndarr[:, 1:]
+            else:
+                I_ndarr.drop(I_ndarr.columns[0], axis=1)
+                if return_distance:
+                    D_ndarr.drop(D_ndarr.columns[0], axis=1)
+
+        return (D_ndarr, I_ndarr) if return_distance else I_ndarr
+
+    def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity'):
+        """
+        Find the k nearest neighbors of column vectors in X and return as
+        a sparse matrix in CSR format.
+
+        Parameters
+        ----------
+        X : array-like (device or host) shape = (n_samples, n_features)
+            Dense matrix (floats or doubles) of shape (n_samples, n_features).
+            Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+            ndarray, cuda array interface compliant array like CuPy
+
+        n_neighbors : Integer
+            Number of neighbors to search. If not provided, the n_neighbors
+            from the model instance is used
+
+        mode : string (default='connectivity')
+            Values in connectivity matrix: 'connectivity' returns the
+            connectivity matrix with ones and zeros, 'distance' returns the
+            edges as the Euclidean distance between points
```

Comment:
`Euclidean` still needs to be changed to something like `as the distances between points with the requested metric`